web-api-db-ctrl/src/postgres.rs

```diff
@@ -0,0 +1,521 @@
+    let mut migrator = sqlx::migrate!("migrations/management");
+    migrator.locking = false;
+    migrator
+});
+
+static TENANT_SCHEMA_MIGRATOR: Lazy<Migrator> = Lazy::new(|| {
+    let mut migrator = sqlx::migrate!("migrations/tenant");
+    migrator.locking = false;
+    migrator
+});
+
+// WARNING: Hardcoding this id to 0 is only okay because we know exactly
+//          which ids are used when. For e.g. sqlx doing so would be a
+//          no-go hence why they derive the id from the db name.
+const MIGRATION_LOCK_ID: i64 = 0;
+
+/// Initializes the DB for multi-tenant usage.
+///
+/// 1. If there is a legacy tenant in public the public schema will
+///    be renamed (and re-owned) to the [`TenantId::random_legacy_tenant_id()`] tenant.
+///
+/// 2. Migrations to the management schema will be run (if needed).
+///
+/// 3. Concurrently for each tenant a migration of their
+///    schema will be run (if needed).
+#[instrument(skip(pool), err)]
+pub(super) async fn initialize(
+    pool: &Pool<Postgres>,
+    enable_legacy_tenant: bool,
+) -> Result<Option<TenantId>, Error> {
+    // Move out to make sure that a pool with a limit of 1 conn doesn't
+    // lead to a dead lock when running tenant migrations. And that we
+    // do release the lock in case of an error.
+    let mut conn = pool.acquire().await?.detach();
+
+    lock_id_until_unlock(&mut conn, MIGRATION_LOCK_ID).await?;
+
+    // WARNING: Many operations here might not be fully transactional.
+    //          Transactions still help with scoping locks and temp.
+    //          session settings.
+    let mut tx = conn.begin().await?;
+
+    info!("running management schema migration");
+    run_migration_in_schema_switch_search_path(
+        &mut tx,
+        &"management".parse()?,
+        &MANAGEMENT_SCHEMA_MIGRATOR,
+    )
+    .await?;
+
+    let legacy_tenant_id = if enable_legacy_tenant {
+        Some(initialize_legacy(&mut tx).await?)
+    } else {
+        None
+    };
+
+    info!("running public schema migration");
+    run_migration_in_schema_switch_search_path(
+        &mut tx,
+        &"public".parse()?,
+        &PUBLIC_SCHEMA_MIGRATOR,
+    )
+    .await?;
+
+    tx.commit().await?;
+
+    // We run this _before_ we release the lock but it will
+    // run concurrently on multiple different connections.
+    //
+    // For this we can have the same guarantees with multi tenant as we
+    // currently have with single tenant.
+    //FIXME: There is a limit to how well this scales.
+    info!("start tenant db schema migrations");
+    let failures = run_all_db_migrations(pool, false).await?;
+
+    unlock_lock_id(&mut conn, MIGRATION_LOCK_ID).await?;
+
+    //TODO we need to decide how to handle partial failure
+    if !failures.is_empty() {
+        for (tenant_id, error) in failures {
+            error!({ %tenant_id, %error }, "migration failed");
+        }
+        bail!("all tenant migrations failed");
+    }
+
+    conn.close().await?;
```

Comment:
❓ shouldn't we move this above l.122 before the error return branching? such that we close the connection no matter what result is returned from the function.